Introduction
============




Background
~~~~~~~~~~

The integration of experimental data to predict phenotypic changes in genome-scale metabolic reconstructions is an active field of research in systems biology. It is imperative to develop software that can generate condition-specific models across a wide range of organisms and data types.

Results
~~~~~~~

We have developed The Algorithm for Simplified Metabolic ANalysIs by Altering Networks and Deducing flux Estimates for VIsuaLization (TASMANIAN DEVIL). There are four separate modules that comprise this software package: gene activity determination, genome-scale metabolic model importation and simplification to reduce network complexity, robust heuristic model building and metabolic flux prediction, and flux visualization from a reference network topology. All modules can be utilized independently or in conjunction with one another. The methods for simplifying models to reduce cofactors and nucleoside phosphate moieties to enhance carbon-centered fluxes are novel. To test the validity of making these changes, publically available experimentally determined flux calculations from aerobic and anaerobic cultures of Saccharomyces cerevisiae were compared to the modeled fluxes generated by our software from gene activity rules derived from paired transcriptomic data sets for these conditions. These simplifications produced significantly more accurate flux predictions compared to fluxes from unmodified models. The software implementation also significantly outperformed a previous release of the model building algorithm in essential reaction determination based off of publically available lethal knockout experiments with paired transcriptomic data sets.

Conclusions
~~~~~~~~~~~

We have designed software that can import the metabolic reconstruction of any model organism and predict and visualize fluxes for a condition based on gene activity rules. We have also proposed and tested a paradigm change to metabolic modeling to decrease the inherent complexity of the models, which demonstrated to be beneficial for flux predictions. Advancements to the model building algorithm enhanced essential reaction determination and flux prediction. Predicting condition-specific in silico fluxes has widespread applications ranging from engineering bacterial and yeast strains for biomass production, to predicting tissue-specific models for drug discovery, to characterizing general metabolic phenotypes for an experiment. TASMANIAN DEVIL has the potential to be adopted by a wide spectrum of researchers using Linux or macOS platforms and is freely available https://github.com/tlawrence3/Tasmanian-Devil.

